lib/rex/sync/ref.rb

```diff
@@ -32,6 +32,10 @@ def ref
     self
   end
 
+  # Prevent duplication of stateful reference counted objects such
+  # as those performed by :replicant Scanner mixin
+  alias :dup :ref
```

Comment:
After syncing up with @jmartin-r7 and @smcintyre-r7 - we're thinking that this won't be the solution to solving the replicant pattern issue that we're running into, i.e. wanting to intentionally share a single reference to a web server across multiple `run_host` calls in a scanner module.

We should be able to land the ldap modules without this particular change, and we can sync back on this discussion afterwards :+1:

Reply:
Replicant patterns aside, I think this highlights a potential bug either way. I think we'll want to start a separate investigation to see where else in framework is calling this object's `dup` function, and find out why. The easiest path for confirming that will be using tracepoints/adding logs here and seeing where/why that's happening.

Reply:
Offering a way forward here, we are thinking `dup` for this object should `raise` and `replicant` would then already keep the original object.  The `raise` would then help point out the other locations this might break so we can address those rather than mask them.

Reply:
Technically speaking, ` intentionally share a single reference to a <pick something> across` is exactly what the Ref mixin does. For unmanaged resources like sockets and the things that use them directly, we kind of need mixin-level implementation of x-thread guards like for instance catching a `:close` when a refcount is > 0. The Ref mixin itself provides a template for how this can be achieved, and it might be worthwhile to look at domain-specific Ref descendants to handle that sort of thing.